datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssignmentStrategy.java

```diff
@@ -0,0 +1,127 @@
+      Optional<Integer> imbalanceThreshold, Optional<Integer> maxPartitionPerTask, boolean enableElasticTaskAssignment,
+      Optional<Integer> partitionsPerTask, Optional<Integer> partitionFullnessFactorPct, Optional<ZkClient> zkClient,
+      String clusterName) {
+    super(maxTasks, imbalanceThreshold, maxPartitionPerTask, enableElasticTaskAssignment, partitionsPerTask,
+        partitionFullnessFactorPct, zkClient, clusterName);
+    _throughputProvider = throughputProvider;
+    _sourceClusterResolver = sourceClusterResolver;
+  }
+
+  @Override
+  public Map<String, Set<DatastreamTask>> assign(List<DatastreamGroup> datastreams, List<String> instances,
+      Map<String, Set<DatastreamTask>> currentAssignment) {
+    return super.assign(datastreams, instances, currentAssignment);
+  }
+
+  @Override
+  public Map<String, Set<DatastreamTask>> assignPartitions(Map<String, Set<DatastreamTask>> currentAssignment,
+      DatastreamGroupPartitionsMetadata datastreamPartitions) {
+    DatastreamGroup datastreamGroup = datastreamPartitions.getDatastreamGroup();
+    String datastreamGroupName = datastreamGroup.getName();
+    HashMap<String, ClusterThroughputInfo> partitionThroughputInfo;
+
+    // Attempting to retrieve partition throughput info with a fallback mechanism to StickyPartitionAssignmentStrategy
+    try {
+      partitionThroughputInfo = fetchPartitionThroughputInfo();
+    } catch (RetriesExhaustedException ex) {
+      LOG.warn("Attempts to fetch partition throughput timed out. Falling back to regular partition assignment strategy");
+      return super.assignPartitions(currentAssignment, datastreamPartitions);
+    }
+
+    LOG.info("Old partition assignment info, assignment: {}", currentAssignment);
+    Validate.isTrue(currentAssignment.size() > 0,
+        "Zero tasks assigned. Retry leader partition assignment.");
+
+    // Calculating task count and assigned partitions
+    int taskCount = getTaskCountForDatastreamGroup(currentAssignment, datastreamGroupName);
+    List<String> assignedPartitions = getAssignedPartitionsForDatastreamGroup(currentAssignment, datastreamGroupName);
+
+    // Elastic task count validation
+    if (getEnableElasticTaskAssignment(datastreamGroup)) {
+      if (assignedPartitions.isEmpty()) {
+        performElasticTaskCountValidation(datastreamPartitions, taskCount);
+      }
+      updateOrRegisterElasticTaskAssignmentMetrics(datastreamPartitions, taskCount);
+    }
+
+    // Calculating unassigned partitions
+    List<String> unassignedPartitions = new ArrayList<>(datastreamPartitions.getPartitions());
+    unassignedPartitions.removeAll(assignedPartitions);
+
+    // Getting a task count estimate based on throughput
+    String clusterName = _sourceClusterResolver.getSourceCluster(datastreamPartitions.getDatastreamGroup());
+    LoadBasedTaskCountEstimator taskCountEstimator = new LoadBasedTaskCountEstimator();
```

Comment:
Should this be an interface and injected in the `LoadBasedPartitionAssignmentStrategy` constructor? Do we plan to have multiple implementations for this, and swap one with another?